SCClassLibrary/Common/Streams/History.sc

```diff
@@ -325,6 +333,29 @@ History { 		// adc 2006, Birmingham; rewrite 2007.
 		^str.clumps(indices.differentiate)
 	}
 
+	*evaluateLineAt { |index| current.evaluateLineAt(index) }
+
+	evaluateLineAt { |index|
+		var line, codeString, result;
+		if (index.isNil) { ^this };
+		line = lines[index];
+		if (line.isNil) { ^this };
+		codeString = line[2];
+		if (codeString.isNil) { ^this };
+
+		try {
+			result = codeString.interpret;
+			"// History code evaluated line: ".postln;
+			codeString.postcs;
+			"// result: ".postln;
+			result.postln;
+		} {
+			"// History code evaluation FAILED for line: ".postln;
```

Comment:
i don't think you need to capitalize "failed" here. also, it might be a good idea to post the error message in addition to the line that failed

Reply:
The scenario is that you play back a history, possibly in edited form, so the current state is not exactly the same, and some lines may fail. 
Then I want to know when a line went wrong, and which one, so this should warn visibly.
So I would prefer changing this to:
```warn("History code evaluation failed for line: ...") ```

As there may be multiple lines that fail, they should not spam the post window with long error posts every 2 seconds. Maybe we want a flag whether to post errors on playback?

Reply:
> As there may be multiple lines that fail, they should not spam the post window with long error posts every 2 seconds.

Wouldn't it be a good idea to stop at the first line that fails? And you can just take the error message, not the whole stack trace. Also, isn't this method also putting out quite a lot of extra information itself anyway? (4 lines per line evaluated)

Reply:
Hm, seems we need two error modes for two use cases of History.play: 
1. historian mode - reconstruct History to replay as accurately as possible 
-> on error, stop and throw, so one can edit history to fix it
2. showtime mode - replay history in segments, maybe in modified orders
-> on error, keep playing, just post short info  - next line might work again.

Reply:
I'm confused, sorry... I'm just talking about `evaluateAtLine`, not play. Are they related in some way? It doesn't seem that `History.play` would ever result in calling this method.

But, I agree that's a good distinction to make. Would you rather do it as two functions or one?

Reply:
Ah yes, you are right, 
History.play does it differently, and should actually also protect against errors.
evaluateLineAt gets used from HistoryGui, and there it is not guaranteed that every line will work - state may not be as needed for the line, esp. if the line is code shared from other players by network. 
ok, will consider what is best here.